addons/controllers/cpi/cpiconfig_utils.go

```diff
@@ -0,0 +1,307 @@
+	switch cpiConfig.Spec.VSphereCPI.Mode {
+	case CPINonParavirtualMode:
+		return r.mapCPIConfigToDataValuesNonParavirtual(ctx, cpiConfig, cluster)
+	case CPIParavirtualMode:
+		return r.mapCPIConfigToDataValuesParavirtual(ctx, cpiConfig, cluster)
+	default:
+		break
+	}
+	return nil, errors.Errorf("Invalid CPI mode %s, must either be %s or %s", cpiConfig.Spec.VSphereCPI.Mode, CPIParavirtualMode, CPINonParavirtualMode)
+}
+
+// getOwnerCluster verifies that the CPIConfig has a cluster as its owner reference,
+// and returns the cluster. It tries to read the cluster name from the CPIConfig's owner reference objects.
+// If not there, we assume the owner cluster and CPIConfig always has the same name.
+func (r *CPIConfigReconciler) getOwnerCluster(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig) (*clusterapiv1beta1.Cluster, error) {
+	cluster := &clusterapiv1beta1.Cluster{}
+	clusterName := cpiConfig.Name
+
+	// retrieve the owner cluster for the CPIConfig object
+	for _, ownerRef := range cpiConfig.GetOwnerReferences() {
+		if strings.EqualFold(ownerRef.Kind, constants.ClusterKind) {
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: cpiConfig.Namespace, Name: clusterName}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' not found", cpiConfig.Namespace, clusterName))
+		}
+		r.Log.Error(err, fmt.Sprintf("Unable to fetch cluster '%s/%s'", cpiConfig.Namespace, clusterName))
+		return nil, err
+	}
+	r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' is successfully found", cpiConfig.Namespace, clusterName))
+	return cluster, nil
+}
+
+// getVSphereCluster gets the VSphereCluster CR for the cluster object
+func (r *CPIConfigReconciler) getVSphereCluster(ctx context.Context, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereCluster, error) {
+	vsphereCluster := &capvv1beta1.VSphereCluster{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}, vsphereCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Errorf("VSphereCluster %s/%s not found", cluster.Namespace, cluster.Name)
+		}
+		return nil, errors.Errorf("VSphereCluster %s/%s could not be fetched, error %v", cluster.Namespace, cluster.Name, err)
+	}
+	return vsphereCluster, nil
+}
+
+func getSecret(ctx context.Context, c client.Client, namespace, name string) (*v1.Secret, error) {
+	secret := &v1.Secret{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Errorf("Secret %s/%s not found", namespace, name)
+		}
+		return nil, errors.Errorf("Secret %s/%s could not be fetched, error %v", namespace, name, err)
+	}
+	return secret, nil
+}
+
+func getUsernameAndPasswordFromSecret(_ context.Context, _ client.Client, s *v1.Secret) (string, string, error) {
```

Comment:
Why we have `_ context.Context, _ client.Client` but they are not used?

Reply:
removed